### DISCUSSION TITLE
$queryRaw doesn't work with single quotes

### DISCUSSION THREAD

**Discussion by bardzh** [prisma/prisma#9991] [Q&A] [answered]
i have a query that goes 

```sql
select count(id), status,  "createdAt"::date from "ProjectLog" where "projectId" = (select id from "Project" where slug = ${id}) and "createdAt" > current_date - interval '${interval} day' group by "createdAt"::date, status;
```
i've also tried user `Prisma.sql` to pass the value inside the quotes but it keeps throwing error that it expected 1 argument but fount 2.

i did not have this issue with prisma `2.20.1`
this issue is only happening on version `3.3.0`

```shell
Query: select count(id) as count, status,  "createdAt"::date from "ProjectLog" where "projectId" = (select id from "Project" where slug = $1) and "createdAt" >  current_date - interval '$2 day' and key notnull group by "createdAt"::date, status
Param: ["main","30"]
PrismaClientKnownRequestError:
Invalid `prisma.queryRaw()` invocation:


  Your raw query had an incorrect number of parameters. Expected: `1`, actual: `2`.
  code: 'P1016',
  clientVersion: '3.3.0',
  meta: { expected: 1, actual: 2 }
}
```
any suggestions ?
i'm not sure if this is a bug.

**Comment by austincrim** [comment]
Hi @bardiaheshmati 👋!

The `$queryRaw` method **must** be called using a [template string](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#queryraw):
```ts
await prisma.$queryRaw`select count(id), status,  "createdAt"::date from "ProjectLog" where "projectId" = (select id from "Project" where slug = ${id}) and "createdAt" > current_date - interval '${interval} day' group by "createdAt"::date, status;`
```

This was a breaking change in our [3.0.1](https://github.com/prisma/prisma/releases/tag/3.0.1) release.

Could you try calling it in that way and see if you still get an error?

**Reply by bardzh** [reply]
it is still breaking. with the same error.
```shell
Invalid `prisma.queryRaw()` invocation:
  Your raw query had an incorrect number of parameters. Expected: `1`, actual: `2`.
    at async Promise.all (index 0)
  code: 'P1016',
  clientVersion: '3.3.0',
  meta: { expected: 1, actual: 2 }
}
```

```shell
Query: select count(id), status,  "createdAt"::date from "ProjectLog" where "projectId" = (select id from "Project" where slug = $1) and "createdAt" > current_date - interval '$2 day' group by "createdAt"::date, status;
Param: ["main",30]
```

i've set `interval` as both `30 <number>` and `'30' <string>`. non worked

**Reply by austincrim** [reply]
Could you try including the word `day` inside the `interval` variable and then removing the single quotes from that part of the query?

```ts
let interval = '30 day'
await prisma.$queryRaw`select count(id), status,  "createdAt"::date from "ProjectLog" where "projectId" = (select id from "Project" where slug = ${id}) and "createdAt" > current_date - interval ${interval} group by "createdAt"::date, status;`

```

**Reply by bardzh** [reply]
I've tried all of these variations. even tried using `Prisma.sql` in once instance.

```js
const interval = '30 day';
prisma.$queryRaw`select count(id), status,  "createdAt"::date from "ProjectLog" where "projectId" = (select id from "Project" where slug = ${id} ) and "createdAt" > current_date - interval ${interval} group by "createdAt"::date, status order by "createdAt"::date asc;`
```
```shell
Query: select count(id), status,  "createdAt"::date from "ProjectLog" where "projectId" = (select id from "Project" where slug = $1 ) and "createdAt" > current_date - interval $2 group by "createdAt"::date, status order by "createdAt"::date asc;
Param: ["main","30 days"]
Query: ROLLBACK
Param: []
PrismaClientKnownRequestError:
Invalid `prisma.queryRaw()` invocation:


  Raw query failed. Code: `42601`. Message: `db error: ERROR: syntax error at or near "$2"`
    at async Promise.all (index 0)
  code: 'P2010',
  clientVersion: '3.3.0',
  meta: {
    code: '42601',
    message: 'db error: ERROR: syntax error at or near "$2"'
  }
}
```

**Reply by austincrim** [reply]
I have been unable to reproduce this exact error locally. Could you provide a minimal reproduction repo so I can continue to investigate? Thanks in advance!

**Reply by janpio** [reply]
ping @bardiaheshmati

**Reply by bardzh** [reply]
apologies, i've been very busy. i didn't get time to bring this out and create a new project to test. but as of now on my current setup it is still throwing the same error. i'll provide you with the project file in a few days time.

**Reply by jschuur** [reply, marked as answer]
See my [comment](https://github.com/prisma/prisma/issues/10051#issuecomment-1002319695) in the issue on how to replicate this

**Reply by bardzh** [reply]
thank you @jschuur , this is exactly where my code is crashing. for now i ended up creating individual ```prisma.$queryRaw``` statements and used those instead. my query only selects a very specific length of records like past month,  past 3 months which i had to create 6 different queries for this;
One note @janpio this issue doesn't happen on MySQL on a similar query.

**Reply by janpio** [reply]
So seems this is PostgreSQL only? I added a label reflecting that to the issue - makes it easier for us to properly reproduce.

**Reply by bardzh** [reply]
@janpio while i was trying to migrate my existing Postgres to MySQL on planetScale, i managed to write the query in a way that doesn't require the variable to goes inside single quote tag like ``` interval - '${day} days' ``` i'm not sure if that's where the issue is or something else is buggy.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
